index.d.ts

```diff
@@ -84,3 +84,26 @@ const ab: Merge<Foo, Bar> = {a: 1, b: 2};
 ```
 */
 export type Merge<FirstType, SecondType> = Omit<FirstType, Extract<keyof FirstType, keyof SecondType>> & SecondType;
+
+/*
+Create a type that requires at least one of the given keys.
+
+@example
+```
+import {RequireAtLeastOne} from 'type-fest';
+
+type SystemMessages = {
+	macos?: string;
+	linux?: string;
+	windows?: string;
+	default?: string;
+};
+
+const messages: RequireAtLeastOne<SystemMessages, 'macos' | 'linux' | 'windows'> = {macos: 'hey', default: 'hello'};
+```
+*/
+export type RequireAtLeastOne<ObjectType, KeysType extends keyof ObjectType = keyof ObjectType> =
+	Omit<ObjectType, KeysType>
+	& {
+		[Key in KeysType]-?: Required<Pick<ObjectType, Key>>
+	}[KeysType];
```

Comment:
I know I'm asking a lot, but would you be able to describe in words how the type works in words? This can be very valuable to users reading this. We could use the opportunity to teach them.